src/condor_io/condor_crypt_aesgcm.cpp

```diff
@@ -0,0 +1,522 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+
+#include "condor_common.h"
+#include "condor_debug.h"
+#include "condor_crypt_aesgcm.h"
+
+#include <algorithm>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+#define IV_SIZE 16
+#define MAC_SIZE 16
+
+unsigned char g_unset_iv[IV_SIZE] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+/*
+void Condor_Crypt_AESGCM::resetState()
+{
+	dprintf(D_NETWORK, "Condor_Crypt_AESGCM::resetState\n");
+	m_conn_crypto_state.reset(new ConnCryptoState());
+	while (!memcmp(m_conn_crypto_state->m_iv_enc.iv, g_unset_iv, IV_SIZE)) {
+		RAND_pseudo_bytes(m_conn_crypto_state->m_iv_enc.iv, IV_SIZE);
+	}
+	memset(m_conn_crypto_state->m_iv_dec.iv, '\0', IV_SIZE);
+}
+*/
+
+// this function is static
+void Condor_Crypt_AESGCM::initState(StreamCryptoState* stream_state)
+{
+	dprintf(D_NETWORK | D_BACKTRACE, "ZKM: ******: Condor_Crypt_AESGCM::initState for %p.\n", stream_state);
+	if (stream_state) {
+		// reset encrypt side
+		RAND_pseudo_bytes(stream_state->m_iv_enc.iv, IV_SIZE);
+		stream_state->m_ctr_enc = 0;
+
+		// reset decrypt side
+		memset(stream_state->m_iv_dec.iv, '\0', IV_SIZE);
+		stream_state->m_ctr_dec = 0;
+	} else {
+		// resetState();
+		ASSERT("stream_state must not be NULL!");
+	}	
+	dprintf(D_NETWORK, "ZKM: ******: Condor_Crypt_AESGCM::initState: Connection count %u.\n", stream_state->m_ctr_conn);
+}
+
+int Condor_Crypt_AESGCM::ciphertext_size_with_cs(int plaintext_size, StreamCryptoState * stream_state) const
+{
+    int ct_sz = plaintext_size;
+    // Authentication tag is an additional 16 bytes; IV is also 16 bytes
+    //
+    // ideally, we would only need to send the IV on the first transmission,
+    // and it could be computed after that based on the message counter:
+    //
+    ct_sz += MAC_SIZE + IV_SIZE * (stream_state->m_ctr_enc ? 0 : 1);
+
+    // however, if we are opting to send it every time so the two sides are
+    // never out of sync as to whether or not they are expecting it, then use
+    // this line instead:
+    //ct_sz += MAC_SIZE + IV_SIZE;
+
+    return ct_sz;
+}
+
+bool Condor_Crypt_AESGCM::encrypt(Condor_Crypto_State *cs,
+                                  const unsigned char *aad,
+                                  int                  aad_len,
+                                  const unsigned char *input,
+                                  int                  input_len, 
+                                  unsigned char *      output, 
+                                  int                  output_buf_len)
+{
+    dprintf(D_NETWORK, "Condor_Crypt_AESGCM::encrypt **********************\n");
+    dprintf(D_NETWORK, "Condor_Crypt_AESGCM::encrypt with %d bytes of input\n", input_len);
+    StreamCryptoState *stream_state = &(cs->m_stream_crypto_state);
+
+        // output length must be aligned to nearest 128-bit block for AES-GCM
```

Comment:
I don't think this comment is true.  Can just remove since we don't do what it says!